### PR TITLE
Handle errors caused by data corruption more gracefully

### PR DESCRIPTION
**Pull request by beorn7**
@juliusv The announced 4th level... (although level 1 is almost done, and level 3 is already removed).
This fixes https://github.com/prometheus/prometheus/issues/877

This requires all the panic calls upon unexpected data to be converted
into errors returned. This pollute the function signatures quite
lot. Well, this is Go...

The ideas behind this are the following:

- panic only if it's a programming error. Data corruptions happen, and
  they are not programming errors.

- If we detect a data corruption, we "quarantine" the series,
  essentially removing it from the database and putting its data into
  a separate directory for forensics.

- Failure during writing to a series file is not considered corruption
  automatically. It will call setDirty, though, so that a
  crashrecovery upon the next restart will commence and check for
  that.

- Series quarantining and setDirty calls are logged and counted in
  metrics, but are hidden from the user of the interfaces in
  interface.go, whith the notable exception of Append(). The reasoning
  is that we treat corruption by removing the corrupted series, i.e. a
  query for it will return no results on its next call anyway, so
  return no results right now. In the case of Append(), we want to
  tell the user that no data has been appended, though.

Minor side effects:

- Now consistently using filepath.* instead of path.*.

- Introduced structured logging where I touched it. This makes things
  less consistent, but a complete change to structured logging would
  be out of scope for this PR.

### CODE REVIEW

**Comment by beorn7** [comment]
@fabxc Could you review this? @juliusv `s review queue is not getting shorter…

**Comment by beorn7** [comment]
BTW: Clunkiness around error returns will be much reduced lates as a side effect of https://github.com/prometheus/prometheus/pull/1466 . I just didn't want to load the ideas of that PR into this one, too.

File: storage/local/storage.go [hunk 1]
```diff
@@ -1291,6 +1322,122 @@ func (s *memorySeriesStorage) calculatePersistenceUrgencyScore() float64 {
...
+		select {
+		case req := <-s.quarantineRequests:
+			s.purgeSeries(req.fp, req.metric, req.reason)
+			log.
+				With("fingerprint", req.fp).
+				With("metric", req.metric).
+				With("reason", req.reason).
+				Warn("Series quarantined.")
+		case <-s.quarantineStopping:
+			log.Info("Series quarantining stopped.")
```
Comment (fabxc) [inline storage/local/storage.go]: A debug log at best.

**Comment by fabxc** [comment]
I'm concerned by the switching to logging errors somewhere in internal functions and not returning errors probably. This practice has caused us troubles in the past. No means of log-spam control, no means of properly handling errors.

The fact that the Storage interface does not expose any errors but returns "sane defaults" is harmful to begin with. We shouldn't extend this any further.

File: storage/local/storage.go (hunk 1 above)
Comment (beorn7) [inline storage/local/storage.go]: This is in parallel to all the other "foobar stopped." messages. We can downgrade them all to debug, but it should happen consistently.

Do you want me to do this mingled with the rest of the changes? In general, I'd prefer a separate change to de-spam logging. (There are more messages to be thought through.)

Comment (fabxc) [inline storage/local/storage.go]: Yes, it's consistent with what we currently have.

I very much agree with Peter here (http://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html) that this should probably be changed.

Makes total sense to do it consistently later.

**Comment by beorn7** [comment]
But… this PR _introduces_ proper error handling. Before, we were either panicking (which could happen on programming errors but not on data corruption on an external device) or the error was just bubbled up to the user and in fact _not_ handled _at all_.

How to handle internal data corruption? The answer this PR provides is: Quarantine the series. Once the error is handled, it doesn't make sense to bubble it up further. Or in other words: If we still bubbled it up to the user _in addition_ to handling it by quarantining, the user would try again and get exactly the answer they are getting in the first try with PR, namely: The series is now gone.

Logging for highly irregular incidents like data corruption is fine. It is for forensics. It doesn't _replace_ handling the error, it _documents_ the handling.

**Comment by fabxc** [comment]
Yes, internally there's totally better error handling.

I was mostly concerned about the removal from the exported bits – I realize that in the cases where we use it, we don't really have an action plan on what to do about it. That doesn't change anything about the interface in my opinion.
"Sane default" on error is equivalent to `res, _ := f()`.

As for logging, yes this should be a rare occurrence but failure likes to come in bulk. If one series corrupts it's unlikely I will even care. When 200k get corrupted I want to actually investigate and then the logs will be of limited use.
I feel like this should be a metric, which can have necessary dimensionality. Maybe log a generic bulk error at most.

File: storage/local/storage.go (hunk 1 above)
Comment (beorn7) [inline storage/local/storage.go]: https://github.com/prometheus/prometheus/issues/1479

**Comment by beorn7** [comment]
Interfaces that are used by a "real" public, e.g. libraries/packages used by other projects, should be stable, and so you might not want to lightheartedly remove a returned error.

However, the storage interface is used solely within this repo, and I strongly believe we should keep it as well tailored to the current need as possible (it's beefy enough already). Refactoring is easy if we are the only users of the interface. Thus: If there is an error we can return to the caller in a meaningful way, sure, go for it. If there is non, let's not pollute the interface with it.

WRT "sane default": I'm not sure what exactly you are referring to. Where does this PR add returning a "sane default"?

WRT logging: Yes, series corruptions may come in bulk, but that's exactly where I want the logs for human inspections. I want to see exactly which files were corrupted. While the logging volume will increase greatly in that case, I really don't see a better way than to log it. Note that there is _also_ a metric that counts the corruptions. So both needs are satisfied. An existing example is the logging during a crash recovery: It's the worst log spam ever, but it's also the most useful one. This flood of log lines is in fact the Prometheus logs I have most often analyzed in detail. One might say: Make it debug logging, so it has to be switched on for forensics. But the problem with crash recover is that you cannot replay it once it has happened. It only happens once, and you cannot switch on debug logging after the fact. I haven't found yet a better solution for the problem, so I want to stick with the least bad solution until a better shows up. The only consolation I can offer is that the log flood only happens under highly irregular circumstances, so it's less bad than "regular" log spam (like the endless "checkpointing" messages, which should go as part of https://github.com/prometheus/prometheus/issues/1479 ).

**Comment by fabxc** [comment]
These (and the crashrecovery) logs sounds like something that would nicely be replaced by a storage investigation tool, then?
Would allow for dry runs, replaying, etc.

As for the interface, we are the only users, yes. Use-case optimization is fine, too.
However, taking that reasoning to degrade best practices in interface design will accumulate more and more hacks over time. I don't believe this is beneficial in the long run.
On the pro side, ignoring the error from the user perspective is consistent with what we have. 

As for "sane defaults" – what would you call `nopIter` otherwise?

**Comment by beorn7** [comment]
WRT crashrecovery: It is happening automatically when a server crashes and gets restarted by the init system (runit or whatever users use). To apply a storage forensics tool, we needed to prevent a restart, and that's not what users would like in the general case. (They want their monitoring server to be up ASAP to continue monitoring.) The logging of quarantined series is a bit less critical as the quarantined series with its error message is visible in the `orphaned` directory. So I could be convinced to have that log at debug level. My concern would be that it will be less visible to the naive user what has happened. But then we might not want to optimize for the naive user…

WRT interface: I consider it bad interface design to return errors that will always be `nil`. Please note that we are not ignoring any errors here. (We are _handling_ them.)

WRT "sane default": The nopIter is not a sane default. It is the truth. The chunk in question is corrupt. We quarantine the series. It is removed from the storage. The appropriate response for iterating over a non-existing series is a nopIter.

**Comment by fabxc** [comment]
Not providing data for something where there should be data is an error.
The corrupted data is handled in some way for the storage – that has no influence whatsoever on what the user expects to see, which is data.

Just like a webserver wouldn't return `200 OK` with a blank page because it handled an error internally. The user still doesn't get data where there should be data.

But this whole discussion isn't worth blocking this chain of PRs right now since it touches more fundamental problems.

:+1:

**Comment by beorn7** [comment]
> Not providing data for something where there should be data is an error.

I don't disagree. I argue that in case of a corrupt series, there should be no data returned to the user. The error has already been handled at the time the result is returned to the user. If we did not quarantine the series, I would agree: We should consistently return an error. However, returning an error, then quarantine, and then return legitimately no data for a repeat of the very same query is inconsistent.

WRT the warning ("partial result"): Sounds good in principle. But then we needed to keep track of all quarantined series to return that warning for each query that involves a quarantined series (see above). Plus, once we see any data corruption, the odds are the storage has way more of them, so essentially all bets are off and every query result is suspicious from now on. Meta-monitoring needs to alert on dirty storage and any number of quarantined series. And we probably should have a big read warning sign on the web status page about quarantined series and dirty state of the storage.

**Pull request merged**